### ISSUE TITLE
Ideas and planning for future improvements

### ISSUE THREAD

**Issue by alexkilpatrickofficial** [alexkilpatrickofficial/Trading_Bot#2]
**Project Improvement Proposal: Enhancing Model Learning Efficiency**

Below is a structured plan that outlines recommended changes to improve training speed, stability, and overall performance within the project. These updates are aimed at making the reinforcement learning pipeline more efficient so the agent can discover profitable strategies on limited hardware.

---

## Overview

Our objectives focus on:
1. **Frequent and Gradual Updates** – Adjusting episode length and training parameters for more frequent weight updates.
2. **Optimized Data Pipeline** – Minimizing overhead by reusing or precomputing heavy features.
3. **Batching and Vectorization** – Reducing Python overhead and improving hardware utilization.
4. **(Optional) Experience Replay** – Integrating replay buffers if off-policy methods are introduced later.

---

### 1. Environment (envs/MultiStockTradingEnv)

**Goal**: Increase update frequency and reduce episode length to deliver more frequent training signals.

#### a. Configurable Maximum Episode Length
- **Where**: `__init__()` method of `MultiStockTradingEnv`.
- **What to Do**:
  1. Add a parameter `max_episode_length` (default, e.g., `5000`).
  2. Initialize an internal counter (e.g., `self.current_step = 0`).
  3. In `step()`, increment `self.current_step` and terminate the episode (set `done = True`) if it reaches `max_episode_length`.

```python
class MultiStockTradingEnv(...):
    def __init__(self, ..., max_episode_length=5000, **kwargs):
        self.max_episode_length = max_episode_length
        self.current_step = 0
        # ... rest of your init code

    def step(self, action):
        # existing step logic
        self.current_step += 1
        if self.current_step >= self.max_episode_length:
            done = True
        # return observation, reward, done, truncated, info
```

- **Why**: Shorter episodes allow for more frequent gradient updates, accelerating convergence.

#### b. (Optional) Vectorized Batch Processing
- **What to Do**:
  - If feasible, modify the environment logic (in `step()` or a custom function) to process a batch of time steps at once.
- **Why**: Reduces Python overhead and speeds up data collection.

---

### 2. Evaluation (evaluation.py)

**Goal**: Implement more frequent PPO updates, organized logging, and optional gradient accumulation.

#### a. Aggregated Logging with Config ID
- **Where**: Near the creation of your PPO model.
- **What to Do**:
  1. Create a helper function (e.g., `get_model_config_id(chromosome)`) that returns a unique string based on the chromosome’s architecture.
  2. Use this string to form a log directory path.
  3. Pass the path to `tensorboard_log` and callbacks.

```python
import hashlib

def get_model_config_id(chromosome):
    config_str = '_'.join(str(x) for x in chromosome)
    config_hash = hashlib.md5(config_str.encode('utf-8')).hexdigest()[:8]
    return f"hybrid_{config_hash}"

config_id = get_model_config_id(validated_chrom)
agg_tb_log_dir = os.path.join("tensorboard", config_id)

model = PPO(
    policy=HybridPolicy,
    env=env,
    verbose=1,
    tensorboard_log=agg_tb_log_dir,
    # etc.
)
```

- **Why**: Ensures logs for each unique model configuration are aggregated in a single location.

#### b. Adjust PPO Hyperparameters
- **Where**: In the `PPO` instantiation.
- **What to Do**:
  - Lower `n_steps` if you want more frequent updates.
  - Adjust `gae_lambda` (often around `0.95` to `0.99`) to find a balance between short- and long-term returns.

```python
model = PPO(
    policy=HybridPolicy,
    env=env,
    n_steps=1024,  # more frequent updates
    gae_lambda=0.95,
    # other parameters
)
```

- **Why**: Frequent updates can help the model learn faster, while `gae_lambda` tunes how much you value immediate vs. future rewards.

#### c. (Optional) Gradient Accumulation
- **What to Do**:
  - If you have limited GPU memory, implement a custom training loop or callback that accumulates gradients over smaller mini-batches.
- **Why**: Simulates training with a larger batch size without requiring additional GPU memory.

---

### 3. Model Architecture (models.py)

**Goal**: Enhance feature extraction for better trading signals.

#### a. ModularBlock
- **Where**: The `ModularBlock` class.
- **What to Do**:
  1. Ensure proper weight initialization (e.g., Kaiming for Conv layers, Xavier for Linear layers).
  2. (Optional) Add skip connections if block input/output shapes match.

- **Why**: Proper initialization and skip connections can stabilize training and improve gradient flow.

#### b. DynamicHybridFeatureExtractor
- **What to Do**:
  1. Double-check all dimension transformations.
  2. Consider logging or printing dimension sizes for debugging.
  3. Evaluate if LSTM/attention layers would help capture longer-term dependencies.

- **Why**: This module is central to your feature extraction; any improvements here can have a significant impact on overall performance.

---

### 4. (Optional) Experience Replay Module

**Goal**: If off-policy methods (e.g., SAC, DDPG) are used, incorporate replay buffers.

- **What to Do**:
  1. Implement a FIFO buffer storing transitions (`(state, action, reward, next_state, done)`).
  2. Sample mini-batches from the buffer during training.
- **Why**: On-policy PPO doesn’t typically use replay, but if you pivot to an off-policy method, a replay buffer can improve sample efficiency.

---

## Final Integration and Testing
1. **Prototype** – Implement each change incrementally and test to ensure no shape mismatches or logic errors.
2. **Extensive Logging** – Use your existing or updated callbacks to monitor training speed, reward signals, gradient norms, and update frequency.
3. **Tune Hyperparameters** – Adjust `max_episode_length`, `n_steps`, `gae_lambda`, and any new architecture parameters based on initial results.
4. **Document** – Update the README or separate docs with details on how to configure max episode length, log directories, etc.
